feat(setup-checklist): add license validity checking logic

erzhtor opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/55665

This PR adds setup checklist license checking logic for cases when the license is 1) invalid/expired 2) or in a free plan 3) or instance exceeds the allowed user count and no 'true-up' is allowed

## Test plan
- Tests added
- Check the diff

## Sreenshots
| Description | After| 
| -- | -- |
| is free plan | <img width="466" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/6717049/cbcfdc90-46ab-4081-af39-4e2af3cf1e44"> |
| exceeds user count | <img width="466" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/6717049/29284159-b42e-4aa3-8bb2-08f9ebe1d146"> | 
| expired | <img width="466" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/6717049/87e855a4-a955-4312-8dce-a35df26ab98e"> | 
| expiring within 7 days | <img width="466" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/6717049/a94cc1e6-4fb4-49e3-ada3-c1e0b4a293b1"> | 

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

sashaostrikov commented on client/web/src/site-admin/setup-checklist/hooks/useSetupChecklist.tsx:
```diff
@@ -24,36 +31,95 @@ export enum SetupChecklistItem {
+        }`
+    }
+
+    if (license?.tags?.includes('plan:free-1')) {
```
Q: Do we need to return a warning in this case? Looks like "yes, we offer free licenses," but at the same time, we show the warning that "it's a free license".

erzhtor replied:
Good point. Updated to have two type of alerts: warning and danger

sashaostrikov reviewed (approved):
I have a question about a free license warning. Apart from that, it looks good!

erzhtor commented on client/web/src/site-admin/setup-checklist/ChecklistNavbarItem.tsx:
```diff
@@ -38,17 +38,29 @@ export const ChecklistNavbarItem: React.FC = () => {
+                                className={classNames('mr-1', feature.needsSetup ? 'text-danger' : 'text-success')}
+                            />
+                        )}
+                        {feature.needsSetup && (
```
Thanks, updated.

erzhtor commented on client/web/src/nav/StatusMessagesNavItem.tsx:
```diff
@@ -161,13 +163,147 @@ const StatusMessagesNavItemEntry: React.FunctionComponent<React.PropsWithChildre
     </div>
 )
 
+const StatusMessageList: React.FC<{
```
FYI: I haven't changed anything here/in this file, just extracted the function to a separate component.

erzhtor commented:
Just FYI, @sashaostrikov, that I've refactored many approaches in this PR, so re-requested your review.

kopancek commented on client/web/src/site-admin/notifications/NotificationsNavbarItem.tsx:
```diff
@@ -0,0 +1,97 @@
+                {name}
+            </div>
+            <div className="pl-2">
+                {/* todo: handle dismissable alerts */}
```
Should we track this todo somewhere?

erzhtor replied:
Yep, thanks. Created follow up https://github.com/sourcegraph/sourcegraph/issues/55860.

kopancek commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -118,6 +144,10 @@ func init() {
 
 	AlertFuncs = append(AlertFuncs, storageLimitReachedAlert)
 
+	AlertFuncs = append(AlertFuncs, freePlanAlert)
+
+	AlertFuncs = append(AlertFuncs, userCountExceededAlert)
```
Nice!

kopancek commented on cmd/frontend/graphqlbackend/site_alerts_test.go:
```diff
@@ -10,6 +10,7 @@ import (
 	"github.com/sourcegraph/sourcegraph/schema"
 )
 
+// todo: update/add new tests
```
I would love to see tests here before approving the PR.

erzhtor replied:
Updated existing and added new tests!

kopancek reviewed (commented):
Overall looks good to me, love the bell icon for notifications. Much more intuitive than the current one 👍 

The only missing thing for me are tests for the new alerts.

erzhtor commented:
> @erzhtor @kopancek I have higher fidelity designs for the setup checklist here: #55835
> I'm thinking we can combine the design in this ticket with the design in that ticket. I'm thinking we don't do a notification. But the errors and warnings here would be displayed in the "Set up" checklist as long as that feature flag is turned on. This would do away with the idea of a global notification option we see here but just like admin onboarding, I think the setup checklist can be a first step towards doing away with the banners and having a single notification center in the navigation in a future iteration. Let me know what you think!

Thank you @meglynst for the review and for designs! 
1. It is actually way easier to implement universal notification center right now (which is in this PR already) rather than adding a dedicated setup checklist, because our global banners are actually are notifications and it is available to all users, it is just admins see additional banners whereas regular users are limited to certain banners. 
2. Also, our global banners, which I reused in this notifications center, can render any html/markdown, so it is also easier to render some danger/warning/info messages with some links in it that will lead to different configuration pages. Most of the banner messages already have good messaging, so by just changing the way how we render banners like in this PR to notifications, we could make a universal notification center that we can use for any purpose, not only setup onboarding. Also, setup is not limited to initial actions, but post-actions when for example they downgrade license but it doesn't include SSO authentication, so we can just show a danger/warning notification stating it. 

So I think, overall notification center could a great multi-purpose thing that could solve not only setup, but post-setup experience (aka observability) by proactively notifying about the instance setup errors/warnings as well as use it for regular user purposes later (perhaps #discuss-job-fair-user-onboarding could use this notification center too).

erzhtor commented:
> Overall looks good to me, love the bell icon for notifications. Much more intuitive than the current one 👍
> 
> The only missing thing for me are tests for the new alerts.

Thanks @kopancek for the review. Addressed comments and updated/added tests!

BolajiOlajide commented on client/web/src/site-admin/notifications/NotificationsNavbarItem.tsx:
```diff
@@ -0,0 +1,96 @@
+    name,
+    items,
+}) => {
+    const hasError = items.find(item => item.type === AlertType.ERROR)
+    const hasWarning = items.find(item => item.type === AlertType.WARNING)
+    const hasInfo = items.find(item => item.type === AlertType.INFO)
```
How large do you think this `items` array can be? It may be efficient to loop through this array once instead of thrice.

```suggestion
	let hasError: boolean = false
	let hasWarning: boolean = false
	let hasInfo: boolean = false
	
	items.forEach(item => {
		switch (item.type) {
			case AlertType.Error:
				hasError = true
				break
			case AlertType.WARNING:
				hasWarning = true
				break
			case AlertType.INFO:
				hasInfo = true
				break
		}
	})
```

erzhtor replied:
Thanks for raising. An array is not large less than 10 for sure, so I would like to leave it as is if you don't mind, for the sake of readability.

BolajiOlajide commented on client/web/src/site-admin/notifications/NotificationsNavbarItem.tsx:
```diff
@@ -0,0 +1,96 @@
+        <div>
+            <div className="mb-2">
+                <Icon
+                    svgPath={hasError || hasWarning || hasInfo ? mdiInformation : mdiCheckCircle}
```
```suggestion
                    svgPath={(hasError || hasWarning || hasInfo) ? mdiInformation : mdiCheckCircle}
```

What do you think about this? Makes it easier ti understand??

erzhtor replied:
It looks like we don't have any Eslint rules for this case, and I am okay with using any of the approaches. Updated.

BolajiOlajide commented on client/web/src/site-admin/notifications/NotificationsNavbarItem.tsx:
```diff
@@ -0,0 +1,96 @@
+                <Icon
+                    svgPath={hasError || hasWarning || hasInfo ? mdiInformation : mdiCheckCircle}
+                    aria-hidden={true}
+                    className={classNames(
+                        styles.icon,
+                        hasError ? styles.danger : hasWarning ? styles.warning : hasInfo ? styles.info : styles.success
+                    )}
```
```suggestion
                    className={classNames(
                        styles.icon,
                        [styles.danger]: hasError,
                        [styles.warning]: hasWarning,
                        [styles.info]: hasInfo
                        [styles.success]: !(hasError || hasWarning || hasInfo)
                    )}
```

This is easier to understand and read. Nested ternaries are hard to read sometimes, it'll be nice to make use of `if` statements by maybe abstracting the use in other places to a function.

BolajiOlajide replied:
I understand that.

It took me about 2-3 minutes to understand the ternary logic, and I fear it's easy to make a mistake, whereas the object key approach is explicit. 
At first glance, I already understand the snippet you shared. I'd go for repeating it it improves readability and maintainability In my opinion.

erzhtor replied:
I see, updated.

BolajiOlajide commented on client/web/src/site-admin/notifications/NotificationsNavbarItem.tsx:
```diff
@@ -0,0 +1,96 @@
+                    {notConfiguredCount > 0 && (
+                        <sup>
+                            <Badge
+                                variant={hasError ? 'danger' : hasWarning ? 'warning' : hasInfo ? 'info' : 'secondary'}
```
Same nested ternary comment as above.

BolajiOlajide reviewed (approved):
I left a couple of nitpicks that'd be nice to have so the code is cleaner and easier to read.
This looks neat, @erzhtor .

erzhtor commented:
@kopancek, @meglynst just FYI I've merged this PR as is as it was growing quite big, and also Milan's other PR was blocked by this. Also, I've created a follow-up https://github.com/sourcegraph/sourcegraph/issues/55875 to address filtering out only setup-related notification items, and create a separate PR.

rrhyne commented:
@erzhtor  we should not combine notifications and setup steps for these reasons:

* If we put setup steps in notifications, the notifications list will quickly fill up and obscure the fact that there are setup items to complete. We want all those items completed to create a healthy instance. 
* Later down the road when we add a new setup item, the admin may have banner blindness to notifications, the new and potentially critical setup step which may improve instance health would have a high chance of being missed

Due to this dynamic, design has segmented the setup steps and notifications into distinct areas of the UI, and differentiated them visually from each other. More context here:

https://www.figma.com/file/B0jm6d60Ct6DVHzMctccdy?node-id=57:2861&mode=design#526389040

unknwon commented:
drive by to say that I really like these new notifications.
